Use kms for s3 storage

matthewborden opened the pull request:
Previously we were using AWS Parameter Store. However after hitting AWS Rate Limiting on Parameter Store and being refused a limit increase we decided to switch the Chamber S3 Backend. However, a major point of difference between the SSM Backend and the S3 Backend is that the SSM Backend supports AWS KMS Keys.

This PR switches from using AES256 Encryption (read the one key stored by all S3 servers) to using KMS Keys for Encryption (read specific keys that you create via AWS KMS Service, AWS Manages the KMS key but you have control over disabling them when they become compromised and restricting access to them via IAM Roles).

Along the way, I ran into the issue that Chamber is caching the keys and values for each service in a __latest.json file. To make the KMS Encryption worthwhile, I changed the __latest.json be a "latest" file per KMS Key. This differs slightly from the behavior of the SSM Store backend when the user doesn't provide the CHAMBER_KMS_KEY_ALIAS environment variable. In the SSM backend, you as the user are able to read any secrets which you have the KMS Key for. However In the S3 Version (with the changes that I've made, you're not able to see secrets encrypted with a KMS Key that you do not have access to).

Currently, the S3 Bucket looks like:

- service-name:
   - __latest.json
   - Key Named JSON File (eg, database_name.json)
   - Key Named JSON File (eg, database_password.json)
   - Key Named JSON File (eg, database_username.json)
- service-name:
   - __latest.json
   - Key Named JSON File (eg, database_name.json)
   - Key Named JSON File (eg, database_password.json)
   - Key Named JSON File (eg, database_username.json)


With these changes:

- service-name:
  - alias_default_parameter_store_key__latest.json
   - Key Named JSON File (eg, database_name.json)
   - Key Named JSON File (eg, database_password.json)
   - Key Named JSON File (eg, database_username.json)
- service-name:
   - alias_default_parameter_store_key__latest.json
   - alias_very_secret_kms_key__latest.json
   - Key Named JSON File (eg, database_name.json)
   - Key Named JSON File (eg, database_password.json)
   - Key Named JSON File (eg, database_username.json)

Implementation Decisions:
- I've chosen to make `chamber list` not show secrets that are encrypted with KMS Keys that you do not have access to. This is because multiple KMS Keys can be used within a single chamber service and some readers will only have one key.
- `chamber write` and `chamber delete` requires CHAMBER_KMS_KEY_ALIAS to be set to the same KMS Key as an existing Chamber secret. This avoids the latest version of a secret being left encrypted in S3 after you move the secret to a new KMS key.

Commit:
Use KMS Encryption instead of AES256

Switch from using AES256 to KMS for storing individual keys and the latest.json file. The SSM Store supports KMS Encryption and allows more segmentation for which keys a user is allowed to read.

Commit:
Store Latest values for a service by KMS Key

Store the latest keys for each service by the KMS key they were written with. Doing so allows for Chamber to use KMS Encryption on the latest file as well as the individual keys in the service.

Commit:
Store KMS Key and Version in __latest.json File

Store the KMS Key and the version in the latest file. This allows us to combine the latest files together (even though there is a __latest.json file per KMS Key).

Commit:
Don't allow writing or deleting with a different KMS Key.

Instead of attempting to migrate between KMS Keys and leaving the first KMS's latest file in an inaccurate state, we ask users to delete and recreate the chamber secret with the new key.

Commit:
Print current KMS key when asking users to delete Chamber secrets

Before writing an existing secret with a new KMS Key, ask the user to delete the existing Chamber secret with the current key. Tell them which KMS key the secret is currently encrypted with.

Commit:
Handle errors attempting to read per KMS key latest file.

When attempting to read the __latest.json for a KMS Key, treat AccessDenied as the file not existing. We do this because the user may not be able to read keys in the service that are encrypted with a KMS Key they don't have access to.

If we get a different kind of error then panic, this isn't my prefered way to handle the error but the S3 pages API doesn't allow to pass back err as a result.

Commit:
Create new S3-KMS Backend

The changes I've made to the __latest.json format (both schema changes and made multiple of these files, one per KMS Key used) are not backwards compadible with the previous implementation of the S3 backend.

Commit:
Add S3-KMS Backend to README.md

Detail how to use it and it's features.

alexf101 commented on store/s3storeKMS.go:
```diff
@@ -0,0 +1,478 @@
+
+	if val, ok := index.Latest[id.Key]; ok {
+		if val.KMSAlias != s.KMSKey() {
+			return errors.New(fmt.Sprintf("You're attempting to write a Chamber key with a different KMS Key to the one it currently uses. Instead delete the existing chamber key using the current KMS key (%s) and then write the secret using the new KMS Key.", val.KMSAlias))
```
I'm not sure I understand this requirement. What if you've lost access to the old key?

matthewborden replied:
> What if you've lost access to the old key?

You wouldn't be able to read/write it anyway without access to the old key.

> I'm not sure I understand this requirement.

Let KMSKeyA be a KMS Key
Let KMSKeyB be a KMS Key

Let SecretA be a Chamber Secret
Let SecretB be a Chamber Secret

Let ChamberOfSecrets be a Chamber Service

User A has access to KMSKeyA & KMSKeyB
User B has access to only KMSKeyA
User C has access to only KMSKeyB

User A writes SecretA with value ABC to ChamberOfSecrets with KMSKeyA
- This creates `kms__KMSKeyA__latest.json` and stores `{ SecretA: "ABC" }`

UserB writes SecretB with value 123 to ChamberOfSecrets with KMSKeyA
- This uses `kms__KMSKeyA__latest.json` and stores `{ SecretA: "ABC", SecretB: 123 }`

UserA attempts to write SecretA with value DEF to ChamberOfSecrets with KMSKeyB
- This uses `kms__KMSKeyB__latest.json` and stores `{ SecretA: "DEF" }`
- `kms__KMSKeyA__latest.json` *still has the value* `{ SecretA: "ABC", SecretB: 123 }`

User B can now see the value of Secret A as "ABC" and User B can see the value as "DEF".

In order for Chamber to migrate the secret over it needs to delete SecretA from `kms__KMSKeyA__latest.json` and write it to `kms__KMSKeyB__latest.json` but the operation that you first asked for was `chamber write`. I don't want a write command to actually delete any data. I want to explicitly delete and then write.

alexf101 commented on store/s3storeKMS.go:
```diff
@@ -0,0 +1,478 @@
+	return s.readObject(path)
+}
+
+func (s *S3KMSStore) puts3raw(path string, contents []byte) error {
```
puts-3-raw? Oh, put-S3-raw. How about camelizing it? `putS3Raw`

matthewborden replied:
This is just copied from S3 Store. I didn't write it.

alexf101 commented on store/s3storeKMS.go:
```diff
@@ -0,0 +1,478 @@
+			if aerr.Code() == "AccessDenied" {
+				// If we're not able to read the latest index for a KMS Key then proceed like it doesn't exist.
+				// We do this because in a chamber secret folder there might be other secrets written with a KMS Key that you don't have access to.
+				return LatestIndexFile{Latest: map[string]LatestValue{}}, nil
```
Have you run `gofmt` over this?

matthewborden replied:
Just did.

alexf101 commented on store/s3storeKMS.go:
```diff
@@ -0,0 +1,478 @@
+			result, err := s.readLatestFile(key_name)
+			if err != nil {
+				// Panic if we encounter an error reading the index (keeping in mind it handles AccessDenied).
+				panic(fmt.Sprintf("Error reading latest index for KMS Key (%s): %s", key_name, err))
```
:(

jeffparsons replied:
:)

jeffparsons replied:
To summarise convo, you could:

- Outside call to `ListObjectsPages` do something like `paginationErr := nil`
- Instead of panicking, set that and pretend you hit the last page so you can get out of the inner thing. (I assume the inner fn closes over whatever outer variables it references?)
- Check after the `ListObjectsPages` call for that `paginationErr` and bubble it up if non-nil

alexf101 commented:
This sort of makes sense? Are there tests or something to check if it works?

Commit:
Run go fmt

matthewborden commented:
> Are there tests or something to check if it works?

Nope, it requires an S3 Bucket and AWS IAM Roles, so we couldn't contribute it upstream. There are no tests for the existing backend and this is also marked as experimental.

I've manually tested it.

Commit:
Use traditional error handling instead of panic.

This seems to be more traditional and I was just working around the lack of err in the AWS ListObjectsPages method.

jeffparsons commented:
LGTM

Commit:
Prefix Latest File Names with __kms

To avoid treating any secret names that start with kms as index files.
